File: 074_Search_a_2D_Matrix/074_1.c
bool searchMatrix(int** matrix, int matrixSize, int* matrixColSize, int target){
    if (matrixSize == 0 || *matrixColSize == 0 || target < matrix[0][0]) return false;
    int m = matrixSize, n = *matrixColSize;
    int row = 0;
    while (target > matrix[row][n - 1]) {
        row++;
        if (row == matrixSize) return false;
    }
    int lo = 0, hi = n;
    while (lo < hi) {
        int mid = lo + (hi - lo) / 2;
        if (target < matrix[row][mid])
            hi = mid;
        else if (target > matrix[row][mid])
            lo = mid + 1;
        else
            return true;
    }
    return false;
}

File: 074_Search_a_2D_Matrix/074_1.py
from typing import List

class Solution:
    def searchMatrix(self, matrix: List[List[int]], target: int) -> bool:
        if not matrix:
            return False
        m, n = len(matrix), len(matrix[0])
        lo, hi = 0, m * n - 1
        while lo <= hi:
            mid = lo + (hi - lo) // 2
            num = matrix[mid // n][mid % n]
            if target < num:
                hi = mid - 1
            elif target > num:
                lo = mid + 1
            else:
                return True
        return False
